### PR TITLE
Signal subscribe in streaming framework and native streaming

### PR DESCRIPTION
**Pull request by NikolaiShipilov**
## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Description:
Adds automatic signal subscribing:

* **#TBBAS-906** Signal/Streaming framework: Automatically subscribing mirrored signals when signal has at least one connection to the InputPort and unsubscribing when the connection count decreases to zero.
* **#TBBAS-905** Native streaming: Subscribing domain signals automatically by client side when value signal is being subscribed. Adding/Removing reader on server side when signal is being subscribed/usubscribed
* Streaming framework refactoring:
  - SignalRemote renamed to MirroredSignalConfig for interfaces and smart pointers
  - IMirroredSignalConfig inherits ISignalConfig
  - Streaming source management methods moved from ISignalConfig to IMirroredSignalConfig
  - Streaming and MirroredSignal implement additional private interfaces
  - Mirrored signal gets initial DataDescriptorChanged EventPacket directly from active streaming source

### CODE REVIEW

**Review by JakaMohorkoDS — COMMENTED** [review]
Some thoughts about the API and gateway devices that I think aren't addressed by this PR.

API:
  - A simple way to receive streamed data or not should be available on a signal (eg. `ISignal::setIsStreamed(bool isStreamed)`)
	- A signal should receive data from a streaming server only if `isStreamed==true` and a connection is formed.
	- Gateway devices might have an exception to the rule (see below)
  
Gateway devices:
  - Gateway devices should only subscribe to signals that either they, or their clients subscribe to.
     - With the current logic, they'll automatically subscribe to all signals, as the server will form input ports.
  - If a gateway device sets a signal to "IsStreamed(false)", it should still be streamed to any client that subscribes to the signal.

File: core/opendaq/streaming/include/opendaq/mirrored_signal_config.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+/*!
+ * @ingroup opendaq_streamings
+ * @addtogroup opendaq_streaming_signal Mirrored Signal
+ * @{
+ */
+
+/*!
+ * @brief Represents configuration interface for mirrored signals. Allows selecting streaming data sources per signal.
+ */
+DECLARE_OPENDAQ_INTERFACE(IMirroredSignalConfig, ISignalConfig)
```
[the remark this answers is not in the document]
Comment (dejancrn) [reply core/opendaq/streaming/include/opendaq/mirrored_signal_config.h:39]: ISignal IMHO should at least have a property something like "Streaming": bool. As discussed, this should do nothing for local signals.

All other functions are fine if they are behind another interface, we just have to make a good "How to guide".

File: docs/tests/test_quick_start.cpp [hunk 2]
```diff
@@ -169,6 +172,7 @@ TEST_F(QuickStartTest, QuickStartAppReaderNativePseudoDevice)
 
     using namespace std::chrono_literals;
     StreamReaderPtr reader = daq::StreamReader<double, uint64_t>(device.getSignals()[0]);
+    docs_test_helpers::waitForSignalSubscription();
```
[the remark this answers is not in the document]
Comment (dejancrn) [reply docs/tests/test_quick_start.cpp:175]: My question is what happens if signal is not yet subscribed and we read data with reader?

**Review by dejancrn — APPROVED** [review]

File: docs/tests/test_quick_start.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply docs/tests/test_quick_start.cpp:175]: From what I can tell we should still have the descriptor before the signal is subscribed? The streaming meta information is separate from the data.

File: core/opendaq/streaming/src/CMakeLists.txt [hunk 3]
```diff
@@ -31,30 +33,32 @@ set(SRC_Cpp streaming_info_impl.cpp
...
 
 set(SRC_PublicHeaders streaming_impl.h
-                      signal_remote_impl.h
+                      mirrored_signal_impl.h
                       streaming_info_factory.h
 )
 
 set(SRC_PrivateHeaders streaming_info_impl.h
+                       streaming_private.h
+                       mirrored_signal_private.h
```
[the remark this answers is not in the document]
Comment (NikolaiShipilov) [reply core/opendaq/streaming/src/CMakeLists.txt]: Fixed by [commit](https://github.com/openDAQ/openDAQ/pull/32/commits/f64752bd891722740111b96fc37e9caa4ffe39f8)

File: docs/tests/test_quick_start.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (NikolaiShipilov) [reply docs/tests/test_quick_start.cpp]: As was discussed client side always has signal descriptors received during connection initialization stage. So when the first listener (e.g. reader) is connected to signal the listener will receive an initial event packet containing these pre-existing descriptors. Subsequently, a subscription request will be initiated, and once the subscription is successfully completed, the server will sent an additional event packet however, this packet will not be forwarded to listeners if the descriptors have not been actually changed.

**Comment by NikolaiShipilov** [comment]
> Some thoughts about the API and gateway devices that I think aren't addressed by this PR.
> 
> API:
> 
> * A simple way to receive streamed data or not should be available on a signal (eg. `ISignal::setIsStreamed(bool isStreamed)`)
>   
>   * A signal should receive data from a streaming server only if `isStreamed==true` and a connection is formed.
>   * Gateway devices might have an exception to the rule (see below)
> 
> Gateway devices:
> 
> * Gateway devices should only subscribe to signals that either they, or their clients subscribe to.
>   
>   * With the current logic, they'll automatically subscribe to all signals, as the server will form input ports.
> * If a gateway device sets a signal to "IsStreamed(false)", it should still be streamed to any client that subscribes to the signal.

With the current logic the gateway device subscribes exclusively to signals utilized by the device itself or signals subscribed by its clients. The mirrored signal owned by a gateway device remain unsubscribed until it have at least one listener. This listener could be an internal gateway entity (such as a function block) or a reader within the gateway streaming server, in case if the signal is subscribed by any of the gateway's clients. In the absence of any client subscribed to the signal, the server reader is just not being created.

However, if the streaming configuration of the mentioned signal is being modified within the gateway device, it will impact how the gateway server streams this signal data to clients. For example, if the signal streaming source is reset within the gateway device, clients that have subscribed will not receive signal data from the gateway server. So, there is a need for a mechanism to separate internal gateway listeners from gateway server readers, ensuring the latter continues to forward signal data independently of the signal's 'isStreamed' status or other streaming settings. 
The implementation of such a mechanism involves making modifications to all handlers within the streaming chain. The details of this mechanism should be thoroughly discussed and carefully designed. I think it could be added at a later stage (out of scope of this PR), along with the addition of the `isStreamed` signal configuration option.

**Review by JakaMohorkoDS — APPROVED** [review]
Lgtm, agreed that we should no go into extreme optimization for gateway devices yet, but add that as necessary. 

The only thing that should be changed is described my comment above (streamed flag on ISignal)

File: core/opendaq/signal/include/opendaq/signal.h [hunk 4]
```diff
@@ -135,6 +135,22 @@ DECLARE_OPENDAQ_INTERFACE(ISignal, IComponent)
...
+    /*!
+     * @brief Returns true if the signal is streamed; false otherwise.
+     * @param[out] streamed True if the signal is streamed; false otherwise.
+     *
+     * A streamed signal receives packets from a streaming server and forwards packets on the signal path.
+     */
+    virtual ErrCode INTERFACE_FUNC getStreamed(Bool* streamed) = 0;
+
+    /*!
+     * @brief Sets the signal to be either streamed or not.
```
Comment (JakaMohorkoDS) [inline core/opendaq/signal/include/opendaq/signal.h:148]: I would add "Setting the "Streamed" flag has no effect if the signal is local to the current Instance. Returns OPENDAQ_IGNORED if that is the case."

**Review by JakaMohorkoDS — APPROVED** [review]

**Pull request merged**